Data.Rmd

```diff
@@ -0,0 +1,567 @@
+This works because environments have reference semantics (whereas more pedestrian R objects, such as atomic vectors, lists, or data frames have value semantics).
+
+[^data-1]: If you don't know much about R environments and what makes them special, a great resource is the [Environments chapter](https://adv-r.hadley.nz/environments.html) of Advanced R.
+
+Consider a package that can store the user's favorite letters or numbers.
+You might start out with code like this in a file below `R/`:
+
+```{r eval = FALSE}
+favorite_letters <- letters[1:3]
+
+#' Report my favorite letters
+#' @export
+mfl <- function() {
+  favorite_letters
+}
+
+#' Change my favorite letters
+#' @export
+set_mfl <- function(l = letters[24:26]) {
+  old <- favorite_letters
+  favorite_letters <<- l
+  invisible(old)
+}
+```
+
+`favorite_letters` is initialized to ("a", "b", "c") when the package is built.
+The user can then inspect `favorite_letters` with `mfl()`, at which point they'll probably want to register *their* favorite letters with `set_mfl()`.
+Note that we've used the super assignment operator `<<-` in `set_mfl()` in the hope that this will reach up into the package environment and modify the internal data object `favorite_letters`.
+But a call to `set_mfl()` fails like so:[^data-2]
+
+[^data-2]: This example will execute without error if you define `favorite_letters`, `mfl()`, and `set_mfl()` in the global workspace and call `set_mfl()` in the console.
+    But this code will fail once `favorite_letters`, `mfl()`, and `set_mfl()` are defined *inside a package*.
+
+```{r, eval = FALSE}
+mfl()
+#> [1] "a" "b" "c"
+
+set_mfl(c("j", "f", "b"))
+#> Error in set_mfl() : 
+#>   cannot change value of locked binding for 'favorite_letters'
+```
+
+Because `favorite_letters` is a regular character vector, modification requires making a copy and rebinding the name `favorite_letters` to this new value.
+And that is what's disallowed: you can't change the binding for objects in the package namespace (well, at least not without trying harder than this).
+Defining `favorite_letters` this way only works if you will never need to modify it.
+
+However, if we maintain state within an internal package environment, we **can** modify objects contained in the environment (and even add completely new objects).
+Here's an alternative implementation that uses an internal environment named "the".
+
+```{r, eval = FALSE}
+the <- new.env(parent = emptyenv())
+the$favorite_letters <- letters[1:3]
+
+#' Report my favorite letters
+#' @export
+mfl2 <- function() {
+  the$favorite_letters
+}
+
+#' Change my favorite letters
+#' @export
+set_mfl2 <- function(l = letters[24:26]) {
+  old <- the$favorite_letters
+  the$favorite_letters <- l
+  invisible(old)
+}
+```
+
+Now a user *can* register their favorite letters:
+
+```{r, eval = FALSE}
+mfl2()
+#> [1] "a" "b" "c"
+
+set_mfl2(c("j", "f", "b"))
+#> [1] "j" "f" "b"
```

Comment:
Am I crazy or should this be `[1] "a" "b" "c"`, i.e. the "old" values?

Reply:
And if it returns `invisible(old)` we shouldn't see it at all?

Reply:
Good catch. This is a copy / paste error. I'm transferring output from a toy package that I've got locally.

Yeah here is the correct chunk here. Will fix.

``` r
mfl2()
#> [1] "a" "b" "c"

set_mfl2(c("j", "f", "b"))

mfl2()
#> [1] "j" "f" "b"
```

Reply:
Yeah I feel like we have quite of bit of interesting "on load" stuff right now. But that it's also a bit too experimental (and getting into the weeds) to work into the book today.